README.md
Syne's Thesia
=============

![Syne's Thesia](/logo_synesthesia.png?raw=true "Syne's Thesia")

>Synesthesia is a perceptual phenomenon in which stimulation of one sensory or cognitive pathway leads to involuntary experiences in a second sensory or cognitive pathway.
[Synesthesia - Wikipedia](https://en.wikipedia.org/wiki/Synesthesia)

*Syne's Thesia* is an external edit of *ElvUI* based on the original *Syne's Edit* of *TukUI*.

![Hunter with all name plates enabled and Recount-profile enabled](/Screenshots/Hunter-Nameplates-FriendlyNameplates-Tooltip-PetBat-RecountStyled.png?raw=true "Hunter with all name plates enabled and Recount-profile enabled")

This isn't 1:1 recreation of the *Syne's Edit*, since there has been ton of new features added and some old features removed from the game over the years compared to the era of the original *Syne's Edit*, but I have tried my best to keep the core features and preserve the style of the original UI by reading the original UI's code like Devil reads the Bible and cross-referencing screenshots and counting pixels for hours. Even with all this work put into this project, I couldn't add all of the features the way they were in the original. I also added few small quality of life -changes here and there that weren't in the original edit, but I felt weren't too invading to the spirit of the original.


File | Game Version | Minimum ElvUI version
--- | --- | ---
`Syne's Thesia 1.11.zip -> Syne's Thesia 1.12.zip` | 9.2.5 | 12.60
`Syne's Thesia 1.9.zip -> Syne's Thesia 1.10.zip` | 9.2.0 | 12.60
`Syne's Thesia 1.7.zip -> Syne's Thesia 1.8.zip` | 9.1.5 | 12.00
`Syne's Thesia 1.3.zip -> Syne's Thesia 1.6.zip` | 9.1.0 | 12.00
`Syne's Thesia 1.2.zip` | 9.0.2 | 12.00
`Syne's Thesia 1.1.zip` | 9.0.1 | 12.00
`Syne's Thesia 1.0.zip` | 8.3 | 11.00


## Features

Most of these are from *Syne's Edit* and few of them are my own edits.

- Installer
  - Should pop-up if/after *ElvUI* install has been completed
  - Create style matching (more or less) profiles for *DBM*, *Recount* and *InFlight*
  - Setup *AddOnSkins* for *DBM* and *Recount*
- Layouts
  - Tank/DPS- and Healer-layouts with all the unit frames styled properly
  - You can change between the styles with slash-commands `/dps` and `/heal`
- Layout User-config
  - Hardcoded layouts only affect some of the options for *Party*, *Raid*, *Raid-40* and *Raid Pet* unit frames but I still wanted you to have as much control as possible over your UI so I made the "Layout User-config" -feature
  - You can type your own changes to the parts of the layout otherwise hardcoded
  - On layout change *Syne's Thesia* will parse them after the hardcoded changes has been applied first
- Unit frames
  - `Low mana` -warning for Player with configurable threshold level
  - Combat Feedback on Player and Target frames
  - You can change the *ElvUI* role icons into *Syne's Edit* -styled colored squares
  - You can select between two different aura styles for Target unit frame
  - Player and Target cast bars can be either embedded to the unit frame or detached to be placed anywhere on the screen
  - While in Healing-layout, you can replace the Party frames with Raid frames by enabling `Grid only -mode`
  - You can change the scale of the Grid styled Raid frames
- Name plates
  - All the name plates are styled, even friendly ones for those who want to enable them outside of the instances
- Data texts
  - 8 new data text panels under the action bars
- Action bars
  - Up to 3 bottom action bars with side action bars (limited to 2 bottom action bars with side action bars enabled)
  - Side action bars at the bottom of the screen
  - Possibility to replace the 12 button bottom actions bars with up to 2 wide action bars with 22 buttons each
  - Possibility to enable extra action bar in the middle of the screen (exact position is user place able)
- Chat
  - Possibility to align the timestamps on the right side of the chat on Right Chat -window
- Miscellaneous
  - Buff reminder if you are missing your own buff while in combat
  - Undress-button in Dress Up Frame
  - Stack items in/to -buttons in Bags and Bank
  - Fix to my personal problem where *ElvUI* and *AddOnSkins* with auto hide embedded AddOns would make *Recount* stay invisible

Also at least bags and bank, data bars, mirror bars, tooltips, threat bar and some of the fonts in the UI have been styled to match the *Syne's Edit*.


## Requirements and installation

*Syne's Thesia* has been developed with *ElvUI* versions `11.27` - `11.372`, with WoW patch 8.3, but I have checked the compatibility and made small fixes with newer WoW patches. You should always use the latest version of *ElvUI* and you can check the release table for latest tested WoW patch.

1. If you haven't completed your *ElvUI* installation yet, go through it first (you can just press `Continue`-button on all the steps and press the `Finished`-button at the last step).
2. Installer should show up after login/*ElvUI* installation.
  - If the installer doesn't pop up automatically, you can launch it from the in game ElvUI-config under the `Syne's Thesia` -section by pressing the `Install`-button.
3. During the installation follow the instructions on the screen and don't be afraid if the UI doesn't look right, everything should fall into right place after the UI reload at the end of the process.

To activate the additional profile-features available for *AddOnSkins*, *DBM*, *Recount* and *InFlight* you might have to change the used profile yourself.

- For *DBM* the installer tries to enable profile automatically.
- For *InFlight* the installer should edit the `Default`-profile which is used for every character if the `Character Specific Options`-setting is turned OFF.
- For *Recount* you have to manually enable the newly created `SynesThesia`-profile either by using the GUI or with command `/recount profile choose SynesThesia`.
- For *AddOnSkins* the embed settings might not work properly until you have reloaded the UI at the end of the installation.

After the UI reload at the end of the install process, you can fine-tune the look of the UI through ElvUI-config. Please note, using the slash-commands to change the layout changes some of the *Party*, *Raid*, *Raid-40* and *Raid Pet* unit frames settings, but you can override those by using the *Layout User-config*.


## Layout User-config

I liked the idea of fast and easy layout changes with slash-command and pre-hardcoded settings, but I still wanted for the user to have almost full control over the UI. My solution to this was this feature where you can set your own edits to the UI and on layout change *Syne's Thesia* will evaluate your changes for the selected layout after applying the hardcoded changes first.

The feature consists of two editboxes, one for Tank/DPS-layout and the other for Healer-layout, under the `Syne's Thesia` -section of ElvUI-config where you can paste your own personalised settings. I know this feature sounds pretty advanced compared to the primitive implementation and isn't exactly newbie-friendly, but I assumed most of the users of the external edits either have some knowledge and understanding how this works or don't care about this feature.

1. First edit the group unit frames to your liking through `UnitFrames`-section under ElvUI-config.
2. Export your profile in "Plugin"-format (`Profiles`-section under ElvUI-config) and find the settings you changed and discard the rest.
3. Paste the changes into the right layout's editbox. If the changes doesn't validate, there is most likely an error, typo or missing character in your paste.
4. If you want to position the frames, you can use the `SY:SetMoverPosition()`-function with the frames Mover.
  - Use it like `:SetPoint()`-method with Mover's global name added as first argument.
5. If the changes validates, reapply the layout with slash-command to confirm the wanted results.
6. Repeat from the beginning if needed.

**N.B.:** If the editbox's content doesn't validate, *Syne's Thesia* won't parse your changes on layout change!

##### Example:

```lua
E.db.unitframe.units.party.buffs.sizeOverride = 44
E.db.unitframe.units.party.debuffs.enable = false
E.db["unitframe"]["units"]["party"]["name"]["text_format"] = "[difficultycolor][level] [namecolor][name:long]"
SY:SetMoverPosition("ElvUF_PartyMover", "TOPLEFT", E.UIParent, "TOPLEFT", 100, -100)
```

**N.B.:** Remember, the layout changes only affects parts of the *Party*, *Raid*, *Raid-40* and *Raid Pet* unit frames settings of the UI. All other changes you made to the UI with ElvUI-config won't be touched when changing layouts and can be omitted!


## Screenshots

Screenshots are taken in the *Syne's Thesia* 1.0 (WoW 8.3), but it should look similar or even better with later versions.

<details>
<summary>Target auras</summary>

###### Rogue with Target auras
![Rogue with Target auras](/Screenshots/Rogue-TargetAuras.png?raw=true "Rogue with Target auras")

###### Rogue with Tank/DPS-layout Raid40 frames with Player and Target auras
![Rogue with Tank/DPS-layout Raid40 frames with Player and Target auras](/Screenshots/Rogue-DPSRaid40-Player&TargetAuras.png?raw=true "Rogue with Tank/DPS-layout Raid40 frames with Player and Target auras")

###### Rogue with Alternative-style on Target auras
![Rogue with Alternative-style on Target auras](/Screenshots/Rogue-AltStyleAuras.png?raw=true "Rogue with Alternative-style on Target auras")
</details>

<details>
<summary>Unit frames</summary>

###### Rogue with Tank/DPS-layout Party and Arena frames
![Rogue with Tank/DPS-layout Party and Arena frames](/Screenshots/Rogue-DPSParty-Arena.png?raw=true "Rogue with Tank/DPS-layout Party and Arena frames")

###### Paladin with Healer-layout Party frames
![Paladin with Healer-layout Party frames](/Screenshots/Paladin-HealerParty.png?raw=true "Paladin with Healer-layout Party frames")

###### Rogue with Tank/DPS-layout Raid and Boss frames
![Rogue with Tank/DPS-layout Raid and Boss frames](/Screenshots/Rogue-DPSRaid-Boss.png?raw=true "Rogue with Tank/DPS-layout Raid and Boss frames")

###### Paladin with Healer-layout Raid frames with Grid-scale 1.0
![Paladin with Healer-layout Raid frames with Grid-scale 1.0](/Screenshots/Paladin-HealerRaid-Scale10.png?raw=true "Paladin with Healer-layout Raid frames with Grid-scale 1.0")

###### Paladin with Healer-layout Raid frames with Grid-scale 1.2
![Paladin with Healer-layout Raid frames with Grid scale 1.2](/Screenshots/Paladin-HealerRaid-Scale12.png?raw=true "Paladin with Healer-layout Raid frames with Grid-scale 1.2")

###### Rogue with Tank/DPS-layout Raid40 frames with Player and Target auras
![Rogue with Tank/DPS-layout Raid40 frames with Player and Target auras](/Screenshots/Rogue-DPSRaid40-Player&TargetAuras.png?raw=true "Rogue with Tank/DPS-layout Raid40 frames with Player and Target auras")
</details>

<details>
<summary>Action, stance and pet bars</summary>

###### Druid with stance bar
![Druid with stance bar](/Screenshots/Druid-StanceBar-Tooltip.png?raw=true "Druid with stance bar")

###### Rogue with two '22 button row' action bars
![Rogue with two '22 button row' action bars](/Screenshots/Rogue-ActionBars2Wide.png?raw=true "Rogue with two '22 button row' action bars")

###### Rogue with three action bars
![Rogue with three action bars](/Screenshots/Rogue-ActionBars3Normal.png?raw=true "Rogue with three action bars")

###### Hunter with pet bar and all name plates and Recount-profile enabled
![Hunter with all name plates enabled and Recount-profile enabled](/Screenshots/Hunter-Nameplates-FriendlyNameplates-Tooltip-PetBat-RecountStyled.png?raw=true "Hunter with all name plates enabled and Recount-profile enabled")
</details>

<details>
<summary>Friendly name plates</summary>

###### Hunter with friendly name plates turned on while targeting NPC
![Hunter with friendly name plates turned on while targeting NPC](/Screenshots/Hunter-FriendlyNamePlates-TargetingNPC.png?raw=true "Hunter with friendly name plates turned on while targeting NPC")

###### Hunter with Friendly name plates while targeting Player
![Hunter with friendly name plates turned on while targeting other Player](/Screenshots/Hunter-FriendlyNamePlates-TargetingPlayer.png?raw=true "Hunter with friendly name plates turned on while targeting other Player")
</details>

<details>
<summary>Cast bars</summary>

###### Rogue with embedded cast bars
![Rogue with embedded cast bars](/Screenshots/Rogue-CastbarsEmbedded.png?raw=true "Rogue with embedded cast bars")

###### Rogue with detached cast bars
![Rogue with detached cast bars](/Screenshots/Rogue-CastbarsDetached.png?raw=true "Rogue with detached cast bars")
</details>

<details>
<summary>Bags and bank</summary>

###### Rogue with bags and bank open
![Rogue with bags and bank open](/Screenshots/Rogue-Bags-Bank-UndressButton.png?raw=true "Rogue with bags and bank open")
</details>


## Issues

If you run into any new issues or have an idea how to better achieve what I'm trying to do, don't hesitate to contact me at [Issues](/../../issues/).

#### Known issues

These are the issues I'm aware of, if you have any idea what causes them or how to fix them, don't hesitate to contact me!

##### New chat-tabs opened while in CombatLockdown causes an error

This bug happens when you are in combat lockdown and receive first (opening a new chat-tab) whisper or BN-whisper from someone. The later whispers received from same user doesn't cause any additional errors, but any whispers from other users will cause the same error to fire for every first whisper.

This bug has probably something to do with improper API-calls inside the WoW Restricted Environment when setting up new chat-tab?

```
1x [ADDON_ACTION_BLOCKED] AddOn 'ElvUI' tried to call the protected function 'RightChatPanel:SetSize()'.
[string "@!BugGrabber\BugGrabber.lua"]:519: in function <!BugGrabber\BugGrabber.lua:519>
[string "=[C]"]: in function `SetSize'
[string "@ElvUI\Core\Toolkit.lua"]:191: in function `Size'
[string "@ElvUI\Modules\chat\Chat-Chat.lua"]:869: in function <ElvUI\Modules\chat\Chat.lua:863>
[string "=[C]"]: in function `PositionChat'
[string "@ElvUI\Modules\chat\Chat-Chat.lua"]:1810: in function <ElvUI\Modules\chat\Chat.lua:1770>
[string "=[C]"]: in function `?'
[string "@AddOnSkins\Libs\Ace3\AceHook-3.0\AceHook-3.0-8.lua"]:103: in function <...Ons\AddOnSkins\Libs\Ace3\AceHook-3.0\AceHook-3.0.lua:100>
[string "=[C]"]: in function `FCF_OpenTemporaryWindow'
[string "@FrameXML\FloatingChatFrame.lua"]:2560: in function <FrameXML\FloatingChatFrame.lua:2549>
```

**Affected versions:** 1.0-


##### Collapsing empty space reserved for auras on Target unit frame

If you use the primary style for auras on unit frames, the UI reserves huge amount of screen space for debuffs pushing the aura bars really high if there are only few debuffs. In an ideal world the AddOn could collapse the empty rows of aura icon or having different aura styles for friendly and hostile targets making the aura bars come way down when all the lines aren't fully populated.

This was one of the reasons I created the alternative-style, but it has its limitations on how many buffs it can show which is a shame when targeting friendly target.

**Affected versions:** 1.0-


## What was left out

These were the features (not present in the original) I at one point thought about adding to the UI, but decided later to leave them out for one reason or another.

##### Ordering debuffs on Target unit frame

Currently only debuffs applied by you are saturated. I would like for the UI also sort debuffs applied by you to the front of the list and have rest of the desaturated debuffs listed after. This might already be possible, but was probably outside of my oUF-editing skills.

##### I had this really sweet idea of Player debuffs...

But it would probably have required ton of custom code and it wasn't in *Syne's Edit* -style so I didn't bother implementing it.

##### *Damn Achievement Spam* -style Achievement spam -jammer

*Damn Achievement Spam* was an AddOn that collected all the achievements gained from the chat in short time window and bunched them into one per different achievement. This reduced the number of lines in chat and guild chat if a raid group gained an achievement. Not implemented, because this was never at the top of my priority list, and the time saved was used to improve other aspects of the UI.


## Thanks

I couldn't have done this edit without the help or "help" (Copy&Paste) and inspiration of the following people:

- Syne (*Syne's Edit*)
- Blazeflack (*CodeNameBlaze* and *ElvUI CustomTweaks*)
- Phanx (CombatText from *oUF_Phanx*)
- Vik and Nefarion (Undress-button)

And of course we shall not forget the hard work and inspiration of:

- Everyone who helped Tukz creating TukUI
- Everyone who helped Elv creating ElvUI

I have also taken look at the code and soaked some inspiration from other TukUI and ElvUI edit creators over the years, but I can't possibly remember all of the names, and it would be unfair to list only few of you all.

I would also like to give special thanks to everyone who has helped me directly or indirectly to improve my skills over the years by providing nice coding tips and practices or even full code snippets over the years at TukUI-, WoWInterface- and CurseForge-forums. If you have ever read these forums, you know the familiar names who pop in giving great advices thread after thread, year after year. And only couple of these tips has been ever given straight to me, reading all the tips and tricks you guys and girls have given to other users has saved me from many troubles later.


---

History or "The Longest Programming Project of My Life"
-------------------------------------------------------

I have always remembered the original *Syne's Edit* with warmth and I was bit lost when Syne stopped updating his UI. When I first contacted Syne in 2014 about this recreation-project, I never would have guessed it would take me nearly **6 years** to get this edit done.

First of all, I had already moved on to *ElvUI*, because the *Syne's Edit* was so out of date it was practically impossible to enjoy the game with patch 4.2.0. I chose *ElvUI* over *TukUI* because it had these quality of life -improvements built into it and it saved me the effort of finding/creating AddOns for many of these small things.


CHANGELOG.md
Syne's Thesia
=============

![Syne's Thesia](/logo_synesthesia.png?raw=true "Syne's Thesia")

>Synesthesia is a perceptual phenomenon in which stimulation of one sensory or cognitive pathway leads to involuntary experiences in a second sensory or cognitive pathway.
[Synesthesia - Wikipedia](https://en.wikipedia.org/wiki/Synesthesia)

*Syne's Thesia* is an external edit of *ElvUI* based on the original *Syne's Edit* of *TukUI*.

---

## Changelog

- 1.12 - Add option to prevent Syne's Thesia from overriding your ElvUI font settings
- 1.11 - 9.2.5 .toc bump
- 1.10 - Fix "valuecolor" being set to RBG instead of RGB when using Class-coloring.
- 1.9 - 9.2.0 .toc bump and fix the installation once again (not an error, but this time ElvUI 12.60 changed things around)
- 1.8 - Fix the Stack-button related errors with Bank
- 1.7 - 9.1.5 .toc bump
- 1.6 - Fix colors getting overwritten during the layout selection and fix the "Loot" -chatframe (installation step for this has been broken probably since 9.0)
- 1.5 - Fix the DT bar related errors during installation
- 1.4 - Fix the error on the "LOW MANA" -text on player frame
- 1.3 - 9.1.0 .toc bump
- 1.2 - 9.0.2 .toc bump
- 1.1 - 9.0.1 .toc bump and ElvUI 12.00 -update
- 1.0 - Initial Release

---

SynesThesia/core/core.lua
--[[----------------------------------------------------------
--	Syne's Thesia, an ElvUI edit by ahak @ tukui.org
--
--	This file contains core functions and updates media
----------------------------------------------------------]]--
local E, L, V, P, G, _ = unpack(ElvUI) --Import: Engine, Locales, PrivateDB, ProfileDB, GlobalDB, Localize Underscore
local SY = E:GetModule("SynesThesia")
local M = E:GetModule("Misc")
local LSM = LibStub("LibSharedMedia-3.0")

--Cache global variables
--Lua functions
local _G = _G
local print, select = print, select
local format = string.format
--WoW API / Variables
local C_ChatBubbles_GetAllChatBubbles = C_ChatBubbles.GetAllChatBubbles

--Global variables that we don't cache, list them here for mikk's FindGlobals script
-- GLOBALS: CHAT_FONT_HEIGHTS, NumberFont_Shadow_Med, WorldFrame

--[[
function SY:Print(msg)
	print('|cff00b3ffSyne\'s Thesia:|r', msg)
end
]]
-- Replaced SY:Print() with better one...
function SY:Print(text, ...)
	if text then
		if text:match("%%[dfqs%d%.]") then
			DEFAULT_CHAT_FRAME:AddMessage("|cff00b3ffSyne\'s Thesia:|r " .. format(text, ...))
		else
			DEFAULT_CHAT_FRAME:AddMessage("|cff00b3ffSyne\'s Thesia:|r " .. strjoin(" ", text, tostringall(...)))
		end
	end
end

--Copied from ElvUI
local function RGBToHex(r, g, b)
	r = r <= 1 and r >= 0 and r or 0
	g = g <= 1 and g >= 0 and g or 0
	b = b <= 1 and b >= 0 and b or 0
	return format("%02x%02x%02x", r*255, g*255, b*255)
end

function SY:ColorStr(str, r, g, b)
	local hex
	local coloredString
	
	if r and g and b then
		hex = RGBToHex(r, g, b)
	else
		hex = RGBToHex(0, 0.7, 1) --Light blue
	end
	
	coloredString = "|cff"..hex..str.."|r"
	return coloredString
end

function SY:SetMoverPosition(mover, point, anchor, secondaryPoint, x, y)
	if not _G[mover] then return end
	local frame = _G[mover]

	frame:ClearAllPoints()
	frame:SetPoint(point, anchor, secondaryPoint, x, y)
	E:SaveMoverPosition(mover)
end

function SY:UpdateMedia()
	--SY.media.chatbubblecolor = E:GetColorTable(E.db.SY.general.chatbubblecolor)
	
	--Timestamp Colors
	--[[
	local text = E.db.SY.chat.textColor
	local bracket = E.db.SY.chat.bracketColor
	SY.media.HexTextColor = RGBToHex(text.r, text.g, text.b)
	SY.media.HexBracketColor = RGBToHex(bracket.r, bracket.g, bracket.b)
	]]
end

local function SetFont(obj, font, size, style, r, g, b, sr, sg, sb, sox, soy)
	--SY:Print("SetFont:", type(obj), font, size, style, r, g, b, sr, sg, sb, sox, soy) -- Debug
	if obj then
		obj:SetFont(font, size, style)
		if sr and sg and sb then obj:SetShadowColor(sr, sg, sb) end
		if sox and soy then obj:SetShadowOffset(sox, soy) end
		if r and g and b then obj:SetTextColor(r, g, b)
		elseif r then obj:SetAlpha(r) end
	end
end

-- Tirisfal
-- Shifring Sand

SY.raidscale = 1
function SY:UpdateBlizzardFonts()
	--if not E.db.SY or not E.db.SY.chat or not E.db.SY.chat.editFonts then return end

	--[[
	--Prepare fonts
	local EDITBOXFONT = LSM:Fetch("font", E.db.SY.chat.editboxFont)
	local EDITBOXFONTSIZE = E.db.SY.chat.editboxFontSize

	--Change minimum allowed chat font size from 12 to 8
	CHAT_FONT_HEIGHTS = {8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20}

	if (not EDITBOXFONT) or (not EDITBOXFONTSIZE) then return end -- UpdateBlizzardFonts() gets called several times during initial login, one of them is a bit too early
	--Change some game fonts and styles
	NumberFont_Shadow_Med:SetFont(EDITBOXFONT, EDITBOXFONTSIZE)
	]]

	local NORMAL	= SY.media.font
	local COMBAT	= SY.media.dmgfont
	local ZONE		= SY.media.zone
	local NUMBER	= SY.media.font	
	local CHAT		= SY.media.chat
	local TIP		= SY.media.tooltip	

	_G.UIDROPDOWNMENU_DEFAULT_TEXT_HEIGHT = 12
	_G.CHAT_FONT_HEIGHTS = { 12, 13, 14, 15, 16, 17, 18, 19, 20 }

	_G.UNIT_NAME_FONT		= NORMAL
	_G.NAMEPLATE_FONT		= NORMAL
	_G.DAMAGE_TEXT_FONT	= COMBAT
	_G.STANDARD_TEXT_FONT	= NORMAL

	-- Base fonts
	SetFont(GameTooltipHeader,					NORMAL,	15*SY.raidscale, "OUTLINE")
	SetFont(NumberFont_OutlineThick_Mono_Small,	NUMBER,	12, "OUTLINE")
	SetFont(NumberFont_Outline_Huge,			NUMBER,	28, "THICKOUTLINE")
	SetFont(NumberFont_Outline_Large,			NUMBER,	15, "OUTLINE")
	SetFont(NumberFont_Outline_Med,				NUMBER,	13, "OUTLINE")
	SetFont(NumberFont_Shadow_Med,				CHAT,	12)
	SetFont(NumberFont_Shadow_Small,			CHAT,	12)
	SetFont(QuestFont,							TIP,	14)
	SetFont(QuestFont_Large,					NORMAL,	14)
	SetFont(SystemFont_Large,					NORMAL,	15)
	SetFont(SystemFont_Med1,					NORMAL,	12)
	SetFont(SystemFont_Med3,					NORMAL,	13)
	SetFont(SystemFont_OutlineThick_Huge2,		NORMAL,	20, "THICKOUTLINE")
	SetFont(SystemFont_Outline_Small,			NUMBER,	12, "OUTLINE")
	SetFont(SystemFont_Shadow_Large,			NORMAL,	15)
	SetFont(SystemFont_Shadow_Med1,				TIP,	12*SY.raidscale, "OUTLINE")
	SetFont(SystemFont_Shadow_Med3,				NORMAL,	13)
	SetFont(SystemFont_Shadow_Outline_Huge2,	NORMAL,	20, "OUTLINE")
	SetFont(SystemFont_Shadow_Small,			TIP,	11*SY.raidscale, "OUTLINE")
	SetFont(SystemFont_Small,					NORMAL,	12)
	SetFont(SystemFont_Tiny,					NORMAL,	12)
	SetFont(Tooltip_Med,						TIP,	12*SY.raidscale, "OUTLINE")
	SetFont(Tooltip_Small,						TIP,	12*SY.raidscale, "OUTLINE")
	SetFont(CombatTextFont,						COMBAT,	100, "OUTLINE") -- number here just increase the font quality.
	SetFont(SystemFont_Shadow_Huge1,			ZONE,	20, "THINOUTLINE")
	SetFont(ZoneTextString,						ZONE,	32, "OUTLINE")
	SetFont(SubZoneTextString,					ZONE,	25, "OUTLINE")
	SetFont(PVPInfoTextString,					ZONE,	22, "THINOUTLINE")
	SetFont(PVPArenaTextString,					ZONE,	22, "THINOUTLINE")
end

function SY.CalculateBarBGHeight()
	local height = 0
	local buttonSize = E.db.actionbar.bar1.buttonsize or 32
	local buttonSpacing = E.db.actionbar.bar1.buttonspacing or 4
	if E.db.SY.actionbar.smallbottom then
		if E.db.SY.actionbar.bottomrows == 1 then
			height = buttonSize + (buttonSpacing * 2)
		elseif E.db.SY.actionbar.bottomrows == 2 or E.db.SY.actionbar.split then
			height = (buttonSize * 2) + (buttonSpacing * 3)
		else
			height = (buttonSize * 3) + (buttonSpacing * 4)
		end
	else
		if E.db.SY.actionbar.bottomrows == 1 then
			height = buttonSize + (buttonSpacing * 2)
		else
			height = (buttonSize * 2) + (buttonSpacing * 3)
		end
	end

	return height
end

local blankTex = LSM:Fetch("border", SY.media.blank)
function SY:CenterGradientH(f, w, h, a1, p, a2, x, y)
	local sh = E:Scale(h)
	local sw = E:Scale(w)
	f:SetFrameLevel(1)
	f:SetHeight(sh)
	f:SetWidth(sw)
	f:SetFrameStrata("BACKGROUND")
	f:SetPoint(a1, p, a2, x, y)	
	f.bg1 = f:CreateTexture(nil, "BACKGROUND")
	f.bg1:SetPoint("TOPRIGHT", f)
	f.bg1:SetPoint("BOTTOMLEFT", f, "BOTTOM")
	--f.bg1:SetTexture(SY.media.blank)
	f.bg1:SetTexture(blankTex)
	f.bg1:SetGradientAlpha("Horizontal", 0, 0, 0, 1, 0, 0, 0, 0)
	f.bg2 = f:CreateTexture(nil, "BACKGROUND")
	f.bg2:SetPoint("TOPLEFT", f)
	f.bg2:SetPoint("BOTTOMRIGHT", f, "BOTTOM")
	--f.bg2:SetTexture(SY.media.blank)
	f.bg2:SetTexture(blankTex)
	f.bg2:SetGradientAlpha("Horizontal", 0, 0, 0, 0, 0, 0, 0, 1)
end

function SY:CenterGradientV(f, w, h, a1, p, a2, x, y)
	local sh = E:Scale(h)
	local sw = E:Scale(w)
	f:SetFrameLevel(1)
	f:SetHeight(sh)
	f:SetWidth(sw)
	f:SetFrameStrata("BACKGROUND")
	f:SetPoint(a1, p, a2, x, y)	
	f.bg1 = f:CreateTexture(nil, "BACKGROUND")
	f.bg1:SetPoint("TOPRIGHT", f)
	f.bg1:SetPoint("BOTTOMLEFT", f, "LEFT")
	--f.bg1:SetTexture(SY.media.blank)
	f.bg1:SetTexture(blankTex)
	f.bg1:SetGradientAlpha("Vertical",  0, 0, 0, 1, 0, 0, 0, 0)
	f.bg2 = f:CreateTexture(nil, "BACKGROUND")
	f.bg2:SetPoint("BOTTOMRIGHT", f)
	f.bg2:SetPoint("TOPLEFT", f, "LEFT")
	--f.bg2:SetTexture(SY.media.blank)
	f.bg2:SetTexture(blankTex)
	f.bg2:SetGradientAlpha("Vertical", 0, 0, 0, 0, 0, 0, 0, 1)
end

function SY:LeftGradient(f, w, h, a1, p, a2, x, y)
	local sh = E:Scale(h)
	local sw = E:Scale(w)
	f:SetFrameLevel(1)
	f:SetHeight(sh)
	f:SetWidth(sw)
	f:SetFrameStrata("BACKGROUND")
	f:SetPoint(a1, p, a2, x, y)	
	f.bg = f:CreateTexture(nil, "BACKGROUND")
	f.bg:SetAllPoints(f)
	--f.bg:SetTexture(SY.media.blank)
	f.bg:SetTexture(blankTex)
	f.bg:SetGradientAlpha("Horizontal", 0, 0, 0, 1, 0, 0, 0, 0)
end

function SY:RightGradient(f, w, h, a1, p, a2, x, y)
	local sh = E:Scale(h)
	local sw = E:Scale(w)
	f:SetFrameLevel(1)
	f:SetHeight(sh)
	f:SetWidth(sw)
	f:SetFrameStrata("BACKGROUND")
	f:SetPoint(a1, p, a2, x, y)	
	f.bg = f:CreateTexture(nil, "BACKGROUND")
	f.bg:SetAllPoints(f)
	--f.bg:SetTexture(SY.media.blank)
	f.bg:SetTexture(blankTex)
	f.bg:SetGradientAlpha("Horizontal", 0, 0, 0, 0, 0, 0, 0, 1)
end

local glowTex = LSM:Fetch("statusbar", SY.media.glowTex)
function SY:CreateShadow(f)
	if f.shadow then return end -- we seriously don't want to create shadow 2 times in a row on the same frame.
	local shadow = CreateFrame("Frame", nil, f, BackdropTemplateMixin and "BackdropTemplate")
	shadow:SetFrameLevel(1)
	shadow:SetFrameStrata(f:GetFrameStrata())
	--[[
	shadow:SetPoint("TOPLEFT", E:Scale(-4), E:Scale(4))
	shadow:SetPoint("BOTTOMLEFT", E:Scale(-4), E:Scale(-4))
	shadow:SetPoint("TOPRIGHT", E:Scale(4), E:Scale(4))
	shadow:SetPoint("BOTTOMRIGHT", E:Scale(4), E:Scale(-4))
	]]
	shadow:SetPoint("TOPLEFT", E:Scale(-3), E:Scale(3))
	shadow:SetPoint("BOTTOMLEFT", E:Scale(-3), E:Scale(-4))
	shadow:SetPoint("TOPRIGHT", E:Scale(3), E:Scale(3))
	shadow:SetPoint("BOTTOMRIGHT", E:Scale(3), E:Scale(-4))
	shadow:SetBackdrop( { 
		edgeFile = glowTex,
		edgeSize = E:Scale(3),
		insets = {
			left = E:Scale(5),
			right = E:Scale(5),
			top = E:Scale(5),
			bottom = E:Scale(5)
		},
	})
	shadow:SetBackdropColor(0, 0, 0, 0)
	shadow:SetBackdropBorderColor(0, 0, 0, 0.5)
	f.shadow = shadow
end

function SY:SetFontString(parent, fontName, fontHeight, fontStyle)
	local fs = parent:CreateFontString(nil, "OVERLAY")
	fs:SetFont(fontName, fontHeight, fontStyle)
	fs:SetJustifyH("LEFT")
	fs:SetShadowColor(0, 0, 0)
	fs:SetShadowOffset(1.25, -1.25)
	return fs
end

function SY:FixFontString(fontString, fontName, fontHeight, fontStyle)
	if not fontString then return end
	if fontName or fontHeight or fontStyle then
		fontString:SetFont(fontName, fontHeight, fontStyle)
	end
	--fontString:SetJustifyH("LEFT")
	fontString:SetShadowColor(0, 0, 0)
	fontString:SetShadowOffset(1.25, -1.25)
end

function SY:StylePanel(f, w, h, a1, p, a2, x, y)
	local sh = E:Scale(h)
	local sw = E:Scale(w)
	f:SetFrameLevel(1)
	f:SetHeight(sh)
	f:SetWidth(sw)
	f:SetFrameStrata("BACKGROUND")
	f:SetPoint(a1, p, a2, x, y)
	f:SetBackdrop({
		bgFile = blankTex,
		edgeFile = blankTex,
		tile = false, tileSize = 0, edgeSize = 1,
		insets = { left = -1, right = -1, top = -1, bottom = -1 }
	})
	f:SetBackdropColor(unpack(SY.media.backdropcolor))
	f:SetBackdropBorderColor(unpack(SY.media.bordercolor))
end

local function GeneralOptions()
	E.Options.args.SY.args.misc = {
		order = 100,
		type = "group",
		name = L["Miscellaneous"],
		get = function(info) return E.db.SY.general[ info[#info] ] end,
		set = function(info, value) E.db.SY.general[ info[#info] ] = value end,
		args = {
			miscHeader = {
				order = 10,
				type = "header",
				name = SY:ColorStr(L["Miscellaneous"]),
			},
			buffreminder = {
				order = 20,
				type = "toggle",
				name = L["Buff Reminder"],
				desc = L["This is now the new Inner Fire warning script for all armors/aspects of a class."],
			},
			recountHack = {
				order = 30,
				type = "toggle",
				name = L["Recount fix"],
				desc = L["This hack should fix the problem of Recount not showing up when it should."],
				set = function(info, value)
					E.db.SY.general[ info[#info] ] = value
					if (value) then
						-- Recount QoL
						if IsAddOnLoaded("Recount") then
							Recount.MainWindow:Show()
						end
					end
				end,
			},
			dontChangeFonts = {
				order = 40,
				type = "toggle",
				name = L["Don't change Fonts"],
				desc = L["Don't override font settings set in ElvUI Config."],
			},
			spacer = {
				order = 100,
				type = "description",
				name = "\n\n\n",
			},
			experimentalGroup = {
				order = 110,
				type = "group",
				name = L["Experimental"],
				guiInline = true,
				args = {
					altPlayerDebuffs = {
						order = 10,
						type = "toggle",
						name = L["Experimental 1"],
						desc = L["Experimental 1 Desc"],
					},
					disclaimerDesc = {
						order = 999,
						type = "description",
						name = L["SY_EXPERIMENTAL_DESC"],
					},
				},
			},
		},
	}
end
SY.configs["general"] = GeneralOptions